File: src/lib/arxiv.ts
import Parser from "rss-parser";

const parser = new Parser();

export interface ArxivPaper {
  title: string;
  link: string;
  pubDate: string;
  source: string;
  snippet: string;
  topic: string;
}

const TOPICS = [
  {
    query: "ti:%22key+information+extraction%22+OR+ti:%22document+understanding%22",
    topic: "Key Information Extraction",
    color: "yellow",
  },
  {
    query: "ti:%22fraud+detection%22+OR+ti:%22anomaly+detection%22+AND+cat:cs.LG",
    topic: "Fraud Detection",
    color: "red",
  },
  {
    query: "ti:%22large+language+model%22+OR+ti:%22language+model%22+AND+cat:cs.CL",
    topic: "LLMs",
    color: "blue",
  },
  {
    query: "ti:%22vision+language+model%22+OR+ti:%22multimodal%22+AND+cat:cs.CV",
    topic: "VLMs",
    color: "purple",
  },
  {
    query: "ti:%22medical%22+OR+ti:%22clinical%22+AND+cat:cs.AI",
    topic: "Medical AI",
    color: "green",
  },
];

export async function fetchArxivPapers(): Promise<ArxivPaper[]> {
  const results = await Promise.allSettled(
    TOPICS.map(async ({ query, topic }) => {
      const url = `https://export.arxiv.org/api/query?search_query=${query}&max_results=10&sortBy=submittedDate&sortOrder=descending`;
      const feed = await parser.parseURL(url);
      return (feed.items || []).map((item) => ({
        title: (item.title || "Untitled").replace(/\s+/g, " ").trim(),
        link: item.link || item.id || "",
        pubDate: item.pubDate || item.isoDate || "",
        source: `arXiv · ${topic}`,
        snippet: (item.contentSnippet || item.summary || "").slice(0, 280).trim(),
        topic,
      }));
    })
  );

  const papers: ArxivPaper[] = [];
  for (const result of results) {
    if (result.status === "fulfilled") papers.push(...result.value);
  }

  return papers.sort(
    (a, b) => new Date(b.pubDate).getTime() - new Date(a.pubDate).getTime()
  );
}

export const TOPIC_COLORS: Record<string, string> = Object.fromEntries(
  TOPICS.map(({ topic, color }) => [topic, color])
);

File: server/app/clients/arxiv.py
import asyncio

import feedparser

from app.schemas.news import ArxivPaper

TOPICS = [
    {
        "query": "abs:%22key+information+extraction%22+OR+abs:%22document+understanding%22+OR+abs:%22information+extraction%22",
        "topic": "Key Information Extraction",
    },
    {
        "query": "abs:%22fraud+detection%22+OR+abs:%22anomaly+detection%22",
        "topic": "Fraud Detection",
    },
    {
        "query": "abs:%22large+language+model%22+AND+cat:cs.CL",
        "topic": "LLMs",
    },
    {
        "query": "abs:%22vision+language+model%22+OR+abs:%22multimodal+model%22+AND+cat:cs.CV",
        "topic": "VLMs",
    },
    {
        "query": "abs:%22medical+imaging%22+OR+abs:%22clinical+nlp%22+OR+abs:%22medical+AI%22",
        "topic": "Medical AI",
    },
]


async def _fetch_topic(query: str, topic: str) -> list[ArxivPaper]:
    url = (
        f"https://export.arxiv.org/api/query?search_query={query}"
        f"&max_results=10&sortBy=submittedDate&sortOrder=descending"
    )
    loop = asyncio.get_event_loop()
    feed = await loop.run_in_executor(None, feedparser.parse, url)
    papers = []
    for entry in feed.entries or []:
        papers.append(
            ArxivPaper(
                title=(entry.get("title", "Untitled") or "").replace("  ", " ").strip(),
                link=entry.get("link") or entry.get("id", ""),
                pub_date=entry.get("published", entry.get("updated", "")),
                source=f"arXiv · {topic}",
                snippet=(entry.get("summary", "") or "")[:280].strip(),
                topic=topic,
            )
        )
    return papers


async def fetch_arxiv_papers() -> list[ArxivPaper]:
    results = await asyncio.gather(
        *[_fetch_topic(t["query"], t["topic"]) for t in TOPICS],
        return_exceptions=True,
    )
    papers: list[ArxivPaper] = []
    for r in results:
        if isinstance(r, list):
            papers.extend(r)
    return sorted(papers, key=lambda x: x.pub_date, reverse=True)
